runtime/interpreter/value.go

```diff
@@ -1841,25 +1841,55 @@ func (v *ArrayValue) Accept(interpreter *Interpreter, visitor Visitor) {
 	})
 }
 
-func (v *ArrayValue) Iterate(interpreter *Interpreter, f func(element Value) (resume bool)) {
-	v.iterate(interpreter, v.array.Iterate, f)
+func (v *ArrayValue) Iterate(
+	interpreter *Interpreter,
+	f func(element Value) (resume bool),
+	locationRange LocationRange,
+) {
+	v.iterate(
+		interpreter,
+		v.array.Iterate,
+		f,
+		locationRange,
+	)
 }
 
-func (v *ArrayValue) IterateLoaded(interpreter *Interpreter, f func(element Value) (resume bool)) {
-	v.iterate(interpreter, v.array.IterateLoadedValues, f)
+func (v *ArrayValue) IterateLoaded(
+	interpreter *Interpreter,
+	f func(element Value) (resume bool),
+	locationRange LocationRange,
+) {
+	v.iterate(
+		interpreter,
+		v.array.IterateLoadedValues,
+		f,
+		locationRange,
+	)
 }
 
 func (v *ArrayValue) iterate(
 	interpreter *Interpreter,
 	atreeIterate func(fn atree.ArrayIterationFunc) error,
 	f func(element Value) (resume bool),
+	locationRange LocationRange,
 ) {
 	iterate := func() {
 		err := atreeIterate(func(element atree.Value) (resume bool, err error) {
 			// atree.Array iteration provides low-level atree.Value,
 			// convert to high-level interpreter.Value
+			elementValue := MustConvertStoredValue(interpreter, element)
+
+			// Each element must be transferred before passing onto the function.
+			transferredElement := elementValue.Transfer(
```

Comment:
Are we sure that "lowering" the transfer from the evaluation of `for` loops into the `iterate` function does not somehow affect the semantics of other uses of the `iterate` function (and its callers)?